### PR TITLE
[Github] Use building LLVM as perf-training for CI container

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adjusts the build process for building the toolchain for the CI container to perform more rigorous perf-training for PGO, particularly building the entirety of LLVM as that is what showed the best results while benchmarking. This patch also splits the job into two stages to avoid timeouts due to the large increase in buildtime. There are a couple other hacks added in here to make things work that we can do away with eventually once we're able to run jobs like this on more powerful self-hosted runners.

### CODE REVIEW

File: .github/workflows/containers/github-action-ci/bootstrap.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/clang/cmake/caches/BOLT-PGO.cmake b/clang/cmake/caches/BOLT-PGO.cmake
+index 1a04ca9a74e5..d092820e4115 100644
+--- a/clang/cmake/caches/BOLT-PGO.cmake
++++ b/clang/cmake/caches/BOLT-PGO.cmake
+@@ -4,6 +4,8 @@ set(CLANG_BOOTSTRAP_TARGETS
+   stage2-clang-bolt
+   stage2-distribution
+   stage2-install-distribution
++  clang
++  lld
```
Comment (tstellar) [inline .github/workflows/containers/github-action-ci/bootstrap.patch:10]: Why is this needed?

Comment (boomanaiden154) [reply .github/workflows/containers/github-action-ci/bootstrap.patch:10]: It's needed so that we don't build more than we need to in the stage2-instrumented build. The default targets build way more in that stage and the generate-profdata target only depends on having clang and lld built.

Comment (tstellar) [reply .github/workflows/containers/github-action-ci/bootstrap.patch:10]: Do you need this because of the way you have split the stages?  If you were to to just run `ninja stage2-clang-bolt` from the start would it still be building a lot of unnecessary targets prior to generating the profile data?  I'm mainly wondering if this is a change that would benefit everyone using the BOLT-PGO cache.

Comment (boomanaiden154) [reply .github/workflows/containers/github-action-ci/bootstrap.patch:10]: Yes. `stage1` takes about an hour to build `stage2-instrumented takes about an hour and a half, generating the profile data takes about four hours, and building the final compiler takes about an hour. I've split the build so that stage 1 and stage2-instrumented are built in the first stage and the profile collection and stage2 are built afterwards. There's no explicit build target (from what I can gather) for building the `stage2-instrumented` compiler without doing profile data collection, so I've just added the targets explicitly here.

It's a hack and I don't think there's that much utility in doing this more broadly. If we wanted to make it less of  hack, we could add an explicit target to build the `stage2-instrumented` dependencies without doing the profile collection, but I think this hack works well enough for now and we should be able to move away from this once we have self-hosted runners setup.

Comment (tstellar) [reply .github/workflows/containers/github-action-ci/bootstrap.patch:10]: OK, thanks for the explanation.

**Review by tstellar — APPROVED** [review]

**Comment by boomanaiden154** [comment]
Looks like there was a transient failure in the stage2 build while loading the container image. I've restarted it and everything seems to be going fine. Hopefully that's not an issue going forward.
